api/watch/plan.go

```diff
@@ -137,6 +141,135 @@ OUTER:
 	return nil
 }
 
+// RunWithClientAndHclogAllDatacenters runs a watch plan using an external client and
+// hclog.Logger instance. The plan queries on all datacenters.
+func (p *Plan) RunWithClientAndHclogAllDatacenters(conf *consulapi.Config, client *consulapi.Client, logger hclog.Logger) error {
+	var watchLogger hclog.Logger
+	if logger == nil {
+		watchLogger = newWatchLogger(nil)
+	} else {
+		watchLogger = logger.Named(watchLoggerName)
+	}
+
+	// Loop until we are canceled
+	failures := 0
+
+	p.mapLastParamVal = make(map[string]BlockingParamVal)
+	p.mapLastResult = make(map[string]interface{})
+OUTER:
+	for !p.shouldStop() {
+		p.client = client
+		blockParamVal := make(map[string]BlockingParamVal)
+		result := make(map[string]interface{})
+		catalog := p.client.Catalog()
+		dcs, err := catalog.Datacenters()
+		if err != nil || len(dcs) == 0 {
+			dcs = append(dcs, "") //This will cause to use default DataCenter if err
+		}
+
+		for _, dc := range dcs {
+			p.Datacenter = dc
+			conf.Address = p.address
+			conf.Datacenter = dc
+			conf.Token = p.Token
+			var clientDC *consulapi.Client
+			clientDC, err = consulapi.NewClient(conf)
+			if err != nil {
+				return fmt.Errorf("Failed to connect to agent: %v", err)
+			}
+			p.client = clientDC
+
+			// Invoke the handler
+			blockParamVal[dc], result[dc], err = p.Watcher(p)
+
+			// Check if we should terminate since the function
+			// could have blocked for a while
+			if p.shouldStop() {
+				break OUTER
+			}
+
+			// Handle an error in the watch function
+			if err != nil {
+				// Perform an exponential backoff
+				failures++
+				if blockParamVal[dc] == nil {
+					p.mapLastParamVal[dc] = nil
+				} else {
+					p.mapLastParamVal[dc] = blockParamVal[dc].Next(p.mapLastParamVal[dc])
+				}
+				retry := retryInterval * time.Duration(failures*failures)
+				if retry > maxBackoffTime {
+					retry = maxBackoffTime
+				}
+				watchLogger.Error("Watch errored", "type", p.Type, "error", err, "retry", retry)
+				select {
+				case <-time.After(retry):
+					continue OUTER
+				case <-p.stopCh:
+					return nil
+				}
+			}
```

Comment:
Couldn't this whole section be extracted into a function and re-used in both `RunWithClientAndHclogAllDatacenters` and `RunWithClientAndHclog`?

Reply:
Initial approach and implementation (if you can see my first commit) was to update the existing function `RunWithClientAndHclog` to handle both cases - single datacenter , all datacenters. 

Though initially it seemed like just introducing a loop for all datacenters, but during the course of development and testing it is found that for 'all datacenters' case requires additional things viz -
- in each watch iteration (main or OUTER loop) , catalog.Datacenters() need to be called to get if any latest dcs list
- new `consulapi.Client` need  to be created  while iterating over dcs, hence additional parameter 'consulapi.Config' need to be passed.

Comparatively there are few lines on err handling only that are similar for both cases.
Also wanted to keep it simple for the case of single datacenter or no datacenter specified , where map variables and iterating over map variables are not required.

Reply:
That's a fair point. It's true extracting it could make it less readable considering the small differences.